frontend/webapp/utils/ldap/forms.py

```diff
@@ -0,0 +1,122 @@
+
+    def __init__(self, *args, **kwargs):
+        """Init new LDAP login form."""
+        super(LDAPLoginForm, self).__init__(*args, **kwargs)
+        self._args = args
+        self._kwargs = kwargs
+        if not self.next.data:
+            self.next.data = request.args.get("next", "")
+        self.remember.default = config_value("DEFAULT_REMEMBER_ME")
+
+    def validate(self):
+        """Validate LDAP logins against AD."""
+        if not super(LDAPLoginForm, self).validate():
+            return False
+
+        if self.email.data.strip() == "":
+            self.email.errors.append(get_message("USERID_NOT_PROVIDED")[0])
+            return False
+
+        if self.password.data.strip() == "":
+            self.password.errors.append(get_message("PASSWORD_NOT_PROVIDED")[0])
+            return False
+
+        try:
+            admin_user = _datastore.get_user(1)
+            if self.email.data == admin_user.username:
+                return self._try_local_auth()
+
+            # first we try authenticating against ldap
+            user_dn, ldap_data = _datastore.query_ldap_user(self.email.data)
+
+            if not _datastore.verify_password(user_dn, self.password.data):
+                self.password.errors.append(get_message("INVALID_PASSWORD")[0])
+                return False
+
+            ldap_email = ldap_data[config_value("LDAP_EMAIL_FIELDNAME")].value
+            password = encrypt_password(self.password.data)
+
+            if _datastore.find_user(email=ldap_email):
+                self.user = _datastore.get_user(ldap_email)
+                # note that this is being stored per user login
+                self.user.password = password
+            else:
+                self.user = _datastore.create_user(
+                    username=ldap_email,
+                    email=ldap_email,
+                    password=password,
+                    active=True,
+                )
+                # need to somehow decide what role they are
+                groups = config_value("LDAP_ADMIN_GROUPS")
+                if any(
+                    [
+                        group in ldap_data[config_value("LDAP_ADMIN_GROUPS_FIELDNAME")]
+                        for group in groups
+                    ]
+                ):
+                    role = _datastore.find_role("admin")
+                else:
+                    role = _datastore.find_role("user")
+                _datastore.add_role_to_user(self.user, role)
+                _datastore.commit()
+        except LDAPSocketOpenError:
+            self.email.errors.append("LDAP Server offline")
+            return self._try_local_auth()
+        except LDAPExceptionError:
+            current_app.artemis_logger.exception("")
+            self.email.errors.append("LDAP Auth failed")
+            return False
+
+        return True
+
+    def _try_local_auth(self):
+        current_app.artemis_logger.info("LDAP failed.. trying local auth")
```

Comment:
this message should not be logged when the superuser tries to log in. Instead, it should state: "Superuser login; using local auth instead of LDAP" or sth like that. The original log message can be used for any other case.

Reply:
this is for any local login not only for superuser

Reply:
yes I am talking about the "LDAP failed" part. When the superuser logs in, it is not the LDAP that "fails" --> we simply bypass it since it is a local admin (superuser).

Reply:
@slowr after you fix this minor logging comment I will approve the PR, seems legit and ready for testing.